dbms/src/Flash/Coprocessor/DAGResponseWriter.h

```diff
@@ -30,6 +30,14 @@ class DAGResponseWriter
     /// prepared with sample block
     virtual void prepare(const Block &){};
     virtual void write(const Block & block) = 0;
+
+    // For async writer, `isReadyForWrite` need to be called before calling `write`.
+    // ```
+    // while (!isReadyForWrite()) {}
+    // write(block);
+    // ```
+    virtual bool isReadyForWrite() const { throw Exception("Unsupport async write"); }
```

Comment:
Temporary out-of-sync is allowed here because `isReadyForWrite` is polled, and the state can be quickly sensed by other writers, and a temporary out-of-sync has little effect.

Reply:
Do you mean if even one writer failed to write the block due to out-of-sync, the data won't loss and query goes well still?

Reply:
> Do you mean if even one writer failed to write the block due to out-of-sync, the data won't loss and query goes well still?

yes, because `ConcurrentIOQueue.nonBlockingWrite` always write block successfully, `isReadyForWrite` is used to prevent too many blocks from being written.

Reply:
Once `writer.isReadyForWrite` returns true, `write.write` must succeed or throwing exception because of writer finished.,